File: python/trustme_secrets/_tmfile.py
"""Reader for the .TM container (format version 2).

The layout is public by design. The only secret is the password that unlocks the
private key held inside it.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from pathlib import Path

MAGIC = b"TMKF"
VERSION = 2


class TrustMeError(Exception):
    """Raised when a key file cannot be read or a secret cannot be fetched."""


@dataclass(frozen=True)
class TmFile:
    app_id: str
    key_id: str
    api_url: str
    public_key: bytes
    salt: bytes
    memory_cost_kib: int
    time_cost: int
    parallelism: int
    nonce: bytes
    ciphertext: bytes
    tag: bytes


class _Cursor:
    def __init__(self, data: bytes) -> None:
        self._data = data
        self._pos = 0

    def take(self, count: int) -> bytes:
        if self._pos + count > len(self._data):
            raise TrustMeError("Key file ended unexpectedly.")
        chunk = self._data[self._pos : self._pos + count]
        self._pos += count
        return chunk

    def u8(self) -> int:
        return self.take(1)[0]

    def u16(self) -> int:
        return int.from_bytes(self.take(2), "big")

    def u32(self) -> int:
        return int.from_bytes(self.take(4), "big")

    def str8(self) -> str:
        return self.take(self.u8()).decode("utf-8")

    def str16(self) -> str:
        return self.take(self.u16()).decode("utf-8")

    def blob16(self) -> bytes:
        return self.take(self.u16())


def load(path: Path) -> TmFile:
    try:
        raw = Path(path).read_bytes()
    except OSError as exc:
        raise TrustMeError(f"Could not read key file {path}: {exc}") from exc

    if len(raw) < 100:
        raise TrustMeError(f"File is too small to be a key file: {path}")

    body, stated = raw[:-32], raw[-32:]
    if not hashlib.sha256(body).digest() == stated:
        raise TrustMeError("Key file is corrupt: checksum does not match.")

    cursor = _Cursor(raw)
    if cursor.take(4) != MAGIC:
        raise TrustMeError(f"Not a TrustMe key file: {path}")

    version = cursor.u8()
    if version != VERSION:
        raise TrustMeError(
            f"Unsupported key file version {version}. "
            "Regenerate this file from the TrustMe console."
        )
    cursor.u8()  # flags

    return TmFile(
        app_id=cursor.str8(),
        key_id=cursor.str8(),
        api_url=cursor.str16(),
        public_key=cursor.blob16(),
        salt=cursor.take(16),
        memory_cost_kib=cursor.u32(),
        time_cost=cursor.u32(),
        parallelism=cursor.u8(),
        nonce=cursor.take(12),
        ciphertext=cursor.blob16(),
        tag=cursor.take(16),
    )

File: javascript/tmfile.js
import { createHash } from "node:crypto";
import { readFile } from "node:fs/promises";

const MAGIC = Buffer.from("TMKF", "ascii");
const VERSION = 2;

export class TrustMeError extends Error {
  constructor(message) {
    super(message);
    this.name = "TrustMeError";
  }
}

class Cursor {
  constructor(data) {
    this.data = data;
    this.pos = 0;
  }
  take(count) {
    if (this.pos + count > this.data.length) {
      throw new TrustMeError("Key file ended unexpectedly.");
    }
    const chunk = this.data.subarray(this.pos, this.pos + count);
    this.pos += count;
    return chunk;
  }
  u8() {
    return this.take(1)[0];
  }
  u16() {
    return this.take(2).readUInt16BE(0);
  }
  u32() {
    return this.take(4).readUInt32BE(0);
  }
  str8() {
    return this.take(this.u8()).toString("utf8");
  }
  str16() {
    return this.take(this.u16()).toString("utf8");
  }
  blob16() {
    return Buffer.from(this.take(this.u16()));
  }
}

// Reader for the .TM container (format version 2). The layout is public by
// design; the only secret is the password that unlocks the key inside it.
export async function loadTmFile(path) {
  let raw;
  try {
    raw = await readFile(path);
  } catch (err) {
    throw new TrustMeError(`Could not read key file ${path}: ${err.message}`);
  }

  if (raw.length < 100) {
    throw new TrustMeError(`File is too small to be a key file: ${path}`);
  }

  const body = raw.subarray(0, raw.length - 32);
  const stated = raw.subarray(raw.length - 32);
  if (!createHash("sha256").update(body).digest().equals(stated)) {
    throw new TrustMeError("Key file is corrupt: checksum does not match.");
  }

  const cursor = new Cursor(raw);
  if (!cursor.take(4).equals(MAGIC)) {
    throw new TrustMeError(`Not a TrustMe key file: ${path}`);
  }

  const version = cursor.u8();
  if (version !== VERSION) {
    throw new TrustMeError(
      `Unsupported key file version ${version}. Regenerate this file from the TrustMe console.`
    );
  }
  cursor.u8(); // flags

  return {
    appId: cursor.str8(),
    keyId: cursor.str8(),
    apiUrl: cursor.str16(),
    publicKey: cursor.blob16(),
    salt: cursor.take(16),
    memoryCostKib: cursor.u32(),
    timeCost: cursor.u32(),
    parallelism: cursor.u8(),
    nonce: Buffer.from(cursor.take(12)),
    ciphertext: cursor.blob16(),
    tag: Buffer.from(cursor.take(16)),
  };
}
